cpp/src/groupby/sort/scan.cpp

```diff
@@ -115,51 +121,91 @@ template <>
 void scan_result_functor::operator()<aggregation::RANK>(aggregation const& agg)
 {
   if (cache.has_result(values, agg)) return;
-  CUDF_EXPECTS(helper.is_presorted(),
-               "Rank aggregate in groupby scan requires the keys to be presorted");
-  auto const order_by = get_grouped_values();
-  CUDF_EXPECTS(!cudf::structs::detail::is_or_has_nested_lists(order_by),
-               "Unsupported list type in grouped rank scan.");
+  // TODO review: this denotes key sorted, but the values are not sorted.
```

Comment:
`  [[nodiscard]] bool is_presorted() const { return keys_are_sorted == sorted::YES; }`
The function definition means, only keys are sorted. But the usage in this code assumes that values are sorted too.
So, I added this comment to get opinion from reviewers.

C++ unit tests for groupby rank scan are written with both sorted keys and sorted values.
Added this line as a comment. It's a pre-condition that happens in Spark (keys and values are pre-sorted before groupby calls). AFAIK, this flag is not used by cuDF python at the moment.

Reply:
So basically the rank aggregations rely on users presorting the values _before_ calling the groupby? I haven't written much groupby code so I don't know the details, could the grouping itself also shuffle the values? I guess at minimum we need to document the requirements for users to sort data somewhere.

Reply:
Spark presorts both keys (pre-sorting keys means, it's already grouped), and values (so, values within groups are also sorted). @mythrocks 

But rank aggregations does not depend on this. Only scan algorithms depend on this. In this PR, groupby rank scan can work both ways (pre-sorted or not).

Reply:
What I'm trying to get at is whether it is possible for a user to invoke this API in an unsafe way that produces incorrect results. It initially sounded to me like the answer was yes: a user who didn't presort the values then called a groupby rank scan would silently get incorrect results. But your last comment seems to indicate otherwise:
> In this PR, groupby rank scan can work both ways (pre-sorted or not).

Could you clarify? I just want to make sure that if there is a way for users to misuse this algorithm by failing to sort we at least have clear documentation about it.

Reply:
Spark users already have input keys and values pre-sorted. This works. 
If the data is not pre-sorted, then the algorithm sorts the data, and proceeds. This also works.